pallets/pallet-bonded-coins/src/lib.rs

```diff
@@ -0,0 +1,1577 @@
+		/// given curve, collateral currency, and bonded currencies. The pool
+		/// will be owned by the origin account.
+		///
+		/// # Parameters
+		/// - `origin`: The origin of the call.
+		/// - `curve`: The curve parameters for the pool.
+		/// - `collateral_id`: The ID of the collateral currency.
+		/// - `currencies`: A bounded vector of token metadata for the bonded
+		///   currencies.
+		/// - `denomination`: The denomination for the bonded currencies.
+		/// - `transferable`: A boolean indicating if the bonded currencies are
+		///   transferable.
+		///
+		/// # Returns
+		/// - `DispatchResult`: The result of the dispatch.
+		///
+		/// # Errors
+		/// - `Error::<T>::InvalidInput`: If the denomination is greater than
+		///   the maximum allowed or if the curve input is invalid.
+		/// - `Error::<T>::Internal`: If the conversion to `BoundedVec` fails.
+		/// - Other errors depending on the types in the config.
+		#[pallet::call_index(0)]
+		#[pallet::weight({
+			let currency_length = currencies.len().saturated_into();
+			let weight_polynomial = T::WeightInfo::create_pool_polynomial(currency_length);
+			let weight_square_root = T::WeightInfo::create_pool_square_root(currency_length);
+			let weight_lmsr = T::WeightInfo::create_pool_lmsr(currency_length);
+			weight_polynomial.max(weight_square_root).max(weight_lmsr)
+		})]
+		pub fn create_pool(
+			origin: OriginFor<T>,
+			curve: CurveInput<CurveParameterInputOf<T>>,
+			collateral_id: CollateralAssetIdOf<T>,
+			currencies: BoundedVec<TokenMetaOf<T>, T::MaxCurrencies>,
+			denomination: u8,
+			transferable: bool,
+			min_operation_balance: u128,
+		) -> DispatchResult {
+			let who = T::PoolCreateOrigin::ensure_origin(origin)?;
+
+			ensure!(denomination <= T::MaxDenomination::get(), Error::<T>::InvalidInput);
+			let checked_curve = curve.try_into().map_err(|_| Error::<T>::InvalidInput)?;
+
+			let currency_length = currencies.len();
+
+			let current_asset_id = NextAssetId::<T>::get();
+
+			let (currency_ids, next_asset_id) = Self::generate_sequential_asset_ids(current_asset_id, currency_length)?;
+
+			let pool_id = T::PoolId::from(currency_ids.blake2_256());
+
+			// Todo: change that.
+			T::DepositCurrency::hold(
+				&T::RuntimeHoldReason::from(HoldReason::Deposit),
+				&who,
+				Self::calculate_pool_deposit(currency_length),
+			)?;
+
+			let pool_account = &pool_id.clone().into();
+
+			// Touch the pool account in order to be able to transfer the collateral
+			// currency to it. This should also verify that the currency actually exists.
+			T::CollateralCurrencies::touch(collateral_id.clone(), pool_account, &who)?;
+
+			currencies
+				.into_iter()
+				.zip(currency_ids.iter())
+				.try_for_each(|(entry, asset_id)| -> DispatchResult {
+					let TokenMeta {
+						min_balance,
+						name,
+						symbol,
+					} = entry;
+
+					T::Fungibles::create(asset_id.clone(), pool_account.to_owned(), false, min_balance)?;
```

Comment:
I agree that asset ID generation should probably be done via a hook, the only problem I can't immediately solve is that it's not clear where the NextAssetId storage would live?

Reply:
> Also, how does this interact with the assets pallet? What if there's already an asset with any of the pool IDs that are generated here? It seems the whole pallet will get stuck and require manual overriding of the asset ID. I think the generation of `N` asset IDs should be delegated to the runtime, which has to make sure there are no conflicts.

The latest versions of the assets pallet have this counter included, but the one we currently use doesn't. So this is something of a stop-gap solution before we can draw on existing functionality. This would, again, require some sort of injectable function that produces the new asset ids. We could maybe already move to an injectable implementation but also have the storage entry still included with this pallet so we can use it in this injectable for now?

Reply:
I could also work on a commit that does this to speed things up if you guys want, or else you can tell me if something from above is unclear.

Reply:
Sorry I took the liberty to hack something together. This is how the pallet would look like: https://github.com/KILTprotocol/kilt-node/commit/f3f8f1d9f9d88baa8cf1a6d92a3b91221caf0132 (needs to be refined). The runtime would then keep track of the last generated ID as a storage value in the new assets pallet deployment. Hence the bonded curves pallet does not need to know anything about asset IDs as such.

Reply:
If we can sneak storage into another pallet like this, then that's indeed a good solution! Let's refine this further

Reply:
Yes as long as it's given a name that won't likely clash with anything else in the future, it should be good to go.

Reply:
complete implementation in https://github.com/KILTprotocol/kilt-node/pull/822